### PR TITLE
Daemon: fix hardcode envoy listener addr

### LINKED ISSUE #17156: Layer 7 protocol visibility failed with annotate io.cilium.proxy-visibility
## Bug report

**General Information**

- Cilium version (run `cilium version`)
root@prod-k8s-server:/home/cilium# cilium version
Client: 1.10.3 4145278 2021-07-15T16:11:03+02:00 go version go1.16.6 linux/amd64
Daemon: 1.10.3 4145278 2021-07-15T16:11:03+02:00 go version go1.16.6 linux/amd64

- Kernel version (run `uname -a`)
root@prod-k8s-server:/home/cilium# uname -a
Linux prod-k8s-server 5.13.8-1.el7.elrepo.x86_64 #1 SMP Tue Aug 3 12:51:23 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux
root@prod-k8s-server:/home/cilium# 

- Orchestration system version in use (e.g. `kubectl version`, ...)
root@10.20.24.36k8s]# kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}

- Link to relevant artifacts (policies, deployments scripts, ...)

Execute steps follow: https://docs.cilium.io/en/v1.10/policy/visibility/, when k8s with cilium startup.

- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```

**How to reproduce the issue**

## setup k8s
**k8s server setup script:**
kubeadm init --kubernetes-version=v1.21.2 \
             --skip-phases=addon/kube-proxy \
             --apiserver-advertise-address=10.20.24.36 \
             --ignore-preflight-errors=Swap

## set up cilium
helm template cilium cilium/cilium  --version=1.10.3 \
     --namespace kube-system \
     --set debug.enabled=true \
     --set tunnel=disabled \
     --set nativeRoutingCIDR=10.0.0.0/8 \
     --set autoDirectNodeRoutes=true \
     --set hostServices.enabled=true \
     --set sockops.enabled=true \
     --set loadBalancer.acceleration=native \
     --set nodePort.enabled=true \
     --set devices="eth0"   \
     --set kubeProxyReplacement=strict \
     --set k8sServiceHost=10.20.24.36 \
     --set k8sServicePort=6443 \
     --set hubble.relay.enabled=true \
     --set hubble.ui.enabled=true \
     --set hubble.metrics.enabled="{dns:query;ignoreAAAA,drop,tcp,flow,icmp,http}" \
     >cilium.yaml

## set annotation

  $ kubectl annotate pod -n kube-system coredns-559cb77956-dkqhz  io.cilium.proxy-visibility="<Egress/53/UDP/DNS>,<Egress/80/TCP/HTTP>" --overwrite
 ## status 

 $ cilium status
    /¯¯\
 /¯¯\__/¯¯\    Cilium:         3 errors
 \__/¯¯\__/    Operator:       1 errors
 /¯¯\__/¯¯\    Hubble:         1 errors
 \__/¯¯\__/    ClusterMesh:    disabled
    \__/

DaemonSet         cilium             Desired: 3, Ready: 2/3, Available: 2/3, Unavailable: 1/3
Deployment        cilium-operator    Desired: 2, Ready: 1/2, Available: 1/2, Unavailable: 1/2
Deployment        hubble-relay       Desired: 1, Unavailable: 1/1
Containers:       cilium             Running: 3
                  cilium-operator    Running: 2
                  hubble-relay       Running: 1
Image versions    cilium             quay.io/cilium/cilium:v1.10.3@sha256:8419531c5d3677158802882bdfe2297915c43f2ebe3649551aaac22de9f6d565: 3
                  cilium-operator    quay.io/cilium/operator-generic:v1.10.3@sha256:337ebf27eae4fbad51cc4baf9110b3ec6753320dd33075bc136e2a1865be5eb5: 2
                  hubble-relay       quay.io/cilium/hubble-relay:v1.10.3@sha256:af8ff09fe374c307356a85b0e0c158359a2e7299f93280151301b7f2fac27339: 1
Errors:           cilium             cilium             1 pods of DaemonSet cilium are not ready
                  cilium             cilium-8d9kb       unable to retrieve cilium status: pods "prod-k8s-node2" not found
                  cilium             cilium-hr4n5       controller endpoint-137-regeneration-recovery is failing since 10s (5x): regeneration recovery failed
                  cilium-operator    cilium-operator    1 pods of Deployment cilium-operator are not ready
                  hubble-relay       hubble-rela
y       1 pods of Deployment hubble-relay are not ready

## detail debug log

```
level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg=256.203045ms bpfWaitForELF="6.576µs" bpfWriteELF="995.001µs" containerID= datapathPolicyRevision=3 desiredPolicyRevision=3 endpointID=2123 error="Error while configuring proxy redirects: context cancelled before waiting for proxy updates: context canceled" identity=18556 ipv4= ipv6= k8sPodName=/ mapSync=15.911895ms policyCalculation="91.471µs" prepareBuild="975.963µs" proxyConfiguration="541.991µs" proxyPolicyCalculation="179.971µs" proxyWaitForAck="6.09µs" reason="retrying regeneration" subsys=endpoint total=277.894896ms waitingForCTClean=654ns waitingForLock="1.644µs"
```

## cilium log file with debug mode

[cilium-log.tar.gz](https://github.com/cilium/cilium/files/6980629/cilium-log.tar.gz)

## Another problem

In this case, if I try to delete the coredns pod with visiblility annotation above, the whole host system of that pod hang, that means I **can't login**. It seems that the network broken. reboot needed.

### PR DESCRIPTION
**Pull request by rock-andy**
Without this patch, NetworkPolicy with Envoy would failed if
AF_INET6 address family not supported, and Envoy would report
error like:

```
xdsDetail="Error adding/updating listener(s) cilium-http-egress:14978:
IPv6 addresses are not supported on this machine: [::]:14978\n"
xdsNonce=196 xdsStreamID=1 xdsTypeURL=type.googleapis.com/envoy.config.listener.v3.Listener
```

Fixes: #17156

Signed-off-by: Zhang Qiang <qiangzhang@qiyi.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #issue-number

```release-note
Envoy configuration is fixed to work also when IPv6 is disabled.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/envoy/server.go [hunk 1]
```diff
@@ -414,7 +419,7 @@ func (s *XDSServer) AddMetricsListener(port uint16, wg *completion.WaitGroup) {
 				Address: &envoy_config_core.Address_SocketAddress{
 					SocketAddress: &envoy_config_core.SocketAddress{
 						Protocol:      envoy_config_core.SocketAddress_TCP,
-						Address:       "::",
+						Address:       listenerAddr,
 						Ipv4Compat:    true,
```
Comment (jrajahalme) [inline pkg/envoy/server.go:423]: I suppose having `Ipv4Compat: true` is harmless when listening on an IPv4 is address?

**Review by jrajahalme — APPROVED** [review]
Looks good, one question below, thanks!

**Comment by jrajahalme** [comment]
Added a release note to the description.

File: pkg/envoy/server.go (hunk 1 above)
Comment (rock-andy) [reply pkg/envoy/server.go:423]: Thanks for your time @jrajahalme . I have just checked the related code of envoy about ipv4Compat, and here's the only usage of socket_address.ipv4_compat, that means IPv4Compat only affect for ipv6 connection. Here's the code:

https://github.com/envoyproxy/envoy/blob/main/source/common/network/utility.cc#L137

**Comment by aanm** [comment]
it looks a little bit suspicious the same test failed in 2 different CI runs. Are we sure this isn't a real failure?

**Pull request merged**